src/withUpgradeExecutor.ts

```diff
@@ -0,0 +1,57 @@
+import {
+  Address,
+  encodeFunctionData as viemEncodeFunctionData,
+  EncodeFunctionDataParameters as ViemEncodeFunctionDataParameters,
+} from 'viem';
+import { GetFunctionName } from './types/utils';
+import { sequencerInboxABI } from './contracts/SequencerInbox';
+import { arbOwnerABI } from './contracts/ArbOwner';
+import { upgradeExecutorEncodeFunctionData } from './upgradeExecutorEncodeFunctionData';
+
+type ABIs = typeof sequencerInboxABI | typeof arbOwnerABI;
+type FunctionName<TAbi extends ABIs> = GetFunctionName<TAbi>;
+
+type EncodeFunctionDataParameters<
+  TAbi extends ABIs,
+  TFunctionName extends FunctionName<TAbi>,
+> = ViemEncodeFunctionDataParameters<TAbi, TFunctionName>;
+
+function encodeFunctionData<TAbi extends ABIs, TFunctionName extends GetFunctionName<TAbi>>({
+  abi,
+  functionName,
+  args,
+}: EncodeFunctionDataParameters<TAbi, TFunctionName>) {
+  return viemEncodeFunctionData({
+    abi,
+    functionName,
+    args,
+  } as unknown as ViemEncodeFunctionDataParameters<TAbi, TFunctionName>);
+}
+
+export function withUpgradeExecutor<TAbi extends ABIs, TFunctionName extends FunctionName<TAbi>>(
+  params: EncodeFunctionDataParameters<TAbi, TFunctionName> & {
+    to: Address;
+    upgradeExecutor: false | Address;
+  },
+) {
+  const { upgradeExecutor } = params;
+  if (!upgradeExecutor) {
+    return {
+      to: params.to,
+      data: encodeFunctionData(params),
+      value: BigInt(0),
+    };
+  }
+
+  return {
+    to: upgradeExecutor,
+    data: upgradeExecutorEncodeFunctionData({
+      functionName: 'executeCall',
```

Comment:
let's also make this configurable but default to `executeCall`

Reply:
Changed (but limited to `execute` or `executeCall`)